{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# * Get target reactions from reference GEMs\n",
    "GEMs: genome-scale metabolic models\n",
    "\n",
    "**1. Download genome files from NCBI DB which is reference strains of target organism.**\n",
    "\n",
    "e.g.,)\n",
    "- Target organism: _Pseudomonas putida_ S12 (Accn. No.: NZ_CP009974.1)\n",
    "- Well-characterized model: _Pseudomonas putida_ KT2440 (Accn. No.: NC_002947.4)\n",
    "- High-quality genome annotation: _Pseudomonas aeruginosa_ PAO1 (Accn. No.: NC_002516.2)\n",
    "\n",
    "\n",
    "**2. Find orthologous using BLASTP-pairwise alignment of 'BLAST+' (standalone BLAST) and EDGAR targeted at each strains's CDSs. (Optional : assemble orthologous data)**\n",
    "\n",
    "- Integrate BLASTP-based orthologous data and EDGAR-based orthologous data. (priority : EDGAR > BLASTP)\n",
    "\n",
    "\n",
    "**3. Download genome-scale metabolic models (GEMs) matched with each strains from model DB which likes to BiGG models.**\n",
    "\n",
    "e.g.,)\n",
    "- _P. putida_ KT2440: iJN1462\n",
    "- _P. aeruginosa_ PAO1: iPAE1146\n",
    "\n",
    "\n",
    "**4. If reference GEMs have old locus tag, change the locus tag of orthologous data to the old locus tag of the GEMs**\n",
    "\n",
    "e.g.,) iJN1462 have old locus tag.\n",
    "\n",
    "\n",
    "**5. Input GEMs with orthologous data. And get reactions from reference GEMs.**\n",
    "\n",
    "- If there are multiple reference models, give priority to the more sophisticated model.\n",
    "- Select the objective function of the draft model from the objective function assigned to the reference GEM.\n",
    "- References GEMs are used to get reactions along with Gene-Protein-Reaction (GPR) relationships.\n",
    "\n",
    "e.g.,) 'a' (gene) is matched to 'A' (reaction). If 'a' is orthologous, 'A' is added to target GEM from references GEMs.\n",
    "\n",
    "- (optional) Leave genes that need to be maintained even if they are not orthologous.\n",
    "\n",
    "e.g.,) Virtual gene of spontaneous reaction\n",
    "\n",
    "- (optional) Set the default oxygen uptake rate for the draft model.\n",
    "\n",
    "\n",
    "\n",
    "# * Dependent packages\n",
    " : COBRApy, Biopython, OpenPyXL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import packages\n",
    "\n",
    "import openpyxl\n",
    "import cobra\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import copy\n",
    "from string import ascii_uppercase\n",
    "\n",
    "# Prepare to create up to 27 dynamic variables\n",
    "apb = ''.join(list(ascii_uppercase))\n",
    "\n",
    "print (apb)\n",
    "print (len(apb))\n",
    "\n",
    "for i in apb: globals()[i] = {} # Generate as many dictionary variables as the number of alphabets.\n",
    "\n",
    "# set PATH\n",
    "path = os.path.abspath(os.path.join(os.getcwd(), '..'))\n",
    "print (path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Prepare to files\n",
    "\n",
    "**1) Genome files**\n",
    "\n",
    "**2) Reference GEMs associated with genome files**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Find orthologous\n",
    "\n",
    "**1) Using BLAST+'s (stand-alone) BLASTP-pairwise alignment.**\n",
    "\n",
    "- Criteria : query coverage > 90% and identity > 90%\n",
    "\n",
    "**2) Using EDGAR (https://edgar3.computational.bio.uni-giessen.de/cgi-bin/edgar_login.cgi?cookie_test=1)**\n",
    "\n",
    "- Union of two orthologous data (priority : EDGAR > BLAST+)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that extracts the locus tag of CDSs from the genome file into a list\n",
    "def extract_cds_lt_in_genome(x):  # x : genome file PATH\n",
    "    \n",
    "    print ('<RUNNING>\\nextract locus tag of CDSs from genome({0})\\n: Extracting CDS for target organism...\\n'.format(x))\n",
    "    \n",
    "    file = next(SeqIO.parse(x, \"genbank\"))\n",
    "\n",
    "    feat = file.features\n",
    "\n",
    "    genome = []\n",
    "    for i in feat:\n",
    "        if i.qualifiers.get(\"pseudo\") == [\"\"]:  # Exclude pseudogene\n",
    "            pass\n",
    "        elif i.type == 'CDS':  # Extract CDS type only\n",
    "            genome.append(i)\n",
    "\n",
    "    cds_lt = []\n",
    "    for i in genome:  # Extract locus tag from each CDS type\n",
    "        cds_lt.extend(i.qualifiers['locus_tag'])\n",
    "\n",
    "    \n",
    "    print ('<DONE>\\nextract_cds_lt_in_genome({0})\\n'.format(x))\n",
    "    \n",
    "    \n",
    "    return cds_lt\n",
    "\n",
    "\n",
    "'''\n",
    "both EDGAR and BLASTP file are input before changing in '.xlsx'\n",
    "\n",
    "BLASTP file`s columns are order to this : Query_ID, Subject_ID, Query_coverage, Identity\n",
    "Query_ID : target organism`s locus tag\n",
    "Subject_ID : counter organism`s locus tag\n",
    "Query_coverage : Query coverage\n",
    "Identity : percent identity\n",
    "\n",
    "BLASTP files are assemble to one file. ('n' = blastp result number,  spreadsheet'n' : blastp_result'n')\n",
    "\n",
    "BLASTP`s criteria : Query coverage > 90% and Identity > 90%\n",
    "\n",
    "BLASTP`s maximum target sequence : 5\n",
    "'''\n",
    "\n",
    "# Convert EDGAR orthologous file to dictionary type (considering paralog)\n",
    "def edgar_orthologous (x):  # x : EDGAR file path\n",
    "\n",
    "    print ('<RUNNING>\\nedgar_orthologous({0})\\n: Extracting EDGAR orthologous file...\\n'.format(x))\n",
    "    \n",
    "    edgar_wb = openpyxl.load_workbook(x)  # load EDGAR EXCEL file\n",
    "    sheet_names = edgar_wb.sheetnames  # Sheet list in EDGAR file\n",
    "    edgar = edgar_wb[sheet_names[0]]  # activated sheet with EDGAR orthologous data\n",
    "    \n",
    "    # if EDGAR 3.0 data, run this script.---------\n",
    "    edgar.delete_rows(1)\n",
    "    \n",
    "    max_row = edgar.max_row # max_row of EDGAR file\n",
    "    \n",
    "    for i, j in enumerate(edgar['1']): # max_column of EDGAR\n",
    "        if bool(j.value) == False:\n",
    "            max_col = i\n",
    "            break\n",
    "    \n",
    "\n",
    "    # col_{} (list type)\n",
    "    # : create dynamic variable and insert value\n",
    "    for i in range(max_col):\n",
    "        globals()['pre_col_{}'.format(apb[i])] = edgar[apb[i]]\n",
    "        globals()['col_{}'.format(apb[i])] = []\n",
    "        for j in range(max_row):\n",
    "            globals()['col_{}'.format(apb[i])].append(globals()['pre_col_{}'.format(apb[i])][j].value.replace(', ,psos,', ','))\n",
    "    \n",
    "\n",
    "    cut_row = col_A.index(' - , -')  # determining upper limit index of row\n",
    "\n",
    "    # remove any parts not related to the header and target organism\n",
    "    for i in range(max_col): globals()['col_{}'.format(apb[i])] = globals()['col_{}'.format(apb[i])][1:cut_row]\n",
    "\n",
    "    # A_vs_{} (dict type)\n",
    "    # : create dynamic variables and match the values between the northologous values, mark as locus_tag (considering paralog)\n",
    "    for i in range(max_col):\n",
    "        if max_col - i - 1 == 0:\n",
    "            break\n",
    "\n",
    "        a = globals()['edgar_A_vs_{}'.format(apb[i + 1])] = {}\n",
    "        for j in range(len(col_A)):  # col_A is target orgasnim\n",
    "            b = globals()['col_{}'.format(apb[i + 1])]\n",
    "            if (',IDENTICAL PARALOGS:,' in col_A[j]) and (',IDENTICAL PARALOGS:,' not in b[j]):  # if col_A element is paralog O, and col_{} element is paralog X, same value (col_{} element) for multiple str keys (col_A element)\n",
    "                lst1 = col_A[j].split(',IDENTICAL PARALOGS:,')\n",
    "                for para1 in lst1: a[para1.split(',')[0]] = b[j].split(',')[0].strip()\n",
    "\n",
    "            elif (',IDENTICAL PARALOGS:,' not in col_A[j]) and (',IDENTICAL PARALOGS:,' in b[j]):  # if col_A element is paralog X, and col_{} element is paralog O, list value (col_{} element) on one str key (col_A element)\n",
    "                lst2 = b[j].split(',IDENTICAL PARALOGS:,')\n",
    "                lst3 = []\n",
    "                for para2 in lst2: lst3.append(para2.split(',')[0].strip())\n",
    "                a[col_A[j].split(',')[0]] = lst3\n",
    "\n",
    "            elif (',IDENTICAL PARALOGS:,' in col_A[j]) and (',IDENTICAL PARALOGS:,' in b[j]):  # if col_A element is paralog O, and col_{} element is paralog O, same list value (col_{} element) for multiple str keys (col_A element)\n",
    "                lst1 = col_A[j].split(',IDENTICAL PARALOGS:,')\n",
    "                lst2 = b[j].split(',IDENTICAL PARALOGS:,')\n",
    "                lst3 = []\n",
    "                for para2 in lst2: lst3.append(para2.split(',')[0].strip())\n",
    "                for para1 in lst1: a[para1.split(',')[0]] = lst3\n",
    "\n",
    "            else: a[col_A[j].split(',')[0]] = b[j].split(',')[0].strip()  # if col_A element is paralog X, and col_{} element is paralog X, one str value (col_{} element) on one str key (col_A element)\n",
    "\n",
    "\n",
    "    for i in range(max_col-1):  # remove any keys that are not included in the CDS' locus_tag in the target organism genome file.\n",
    "        a = globals()['edgar_A_vs_{}'.format(apb[i + 1])]\n",
    "        a_keys = list(a.keys())\n",
    "        for j in a_keys:\n",
    "            if j not in target_cds_lt:\n",
    "                del a[j]\n",
    "\n",
    "\n",
    "    result = []\n",
    "    for i in range(max_col-1):  # combine in result (list type)\n",
    "        a = globals()['edgar_A_vs_{}'.format(apb[i + 1])]\n",
    "        result.append(a)\n",
    "\n",
    "    \n",
    "    print ('<DONE>\\nedgar_orthologous({0})\\n'.format(x))\n",
    "    \n",
    "\n",
    "\n",
    "# Convert BLAST+ orthologous file to dictionary type (considering paralog)\n",
    "def blast_orthologous (x):  # x : BLAST+ file path\n",
    "    \n",
    "    print ('<RUNNING>\\nblastp_orthologous({0})\\n: Extracting BLAST+ orthologous file...\\n'.format(x))\n",
    "    \n",
    "    blast_wb = openpyxl.load_workbook(x)\n",
    "    sheet_names = blast_wb.sheetnames\n",
    "\n",
    "    serial = 1\n",
    "    for sheet in sheet_names:  # Output BLAST results sequentially and process them\n",
    "        blast = blast_wb[sheet]\n",
    "\n",
    "        q_id_pre = blast['A']  # Change Query_ID, Subject_ID, Query_coverage, Identity to list format\n",
    "        q_id_pre = q_id_pre[1:]\n",
    "        q_id = []\n",
    "        for i in q_id_pre: q_id.append(i.value.split(',')[0])\n",
    "        s_id_pre = blast['B']\n",
    "        s_id_pre = s_id_pre[1:]\n",
    "        s_id = []\n",
    "        for i in s_id_pre: s_id.append(i.value.split(',')[0])\n",
    "        q_cov_pre = blast['C']\n",
    "        q_cov_pre = q_cov_pre[1:]\n",
    "        q_cov = []\n",
    "        for i in q_cov_pre: q_cov.append(i.value)\n",
    "        ident_pre = blast['D']\n",
    "        ident_pre = ident_pre[1:]\n",
    "        ident = []\n",
    "        for i in ident_pre: ident.append(i.value)\n",
    "\n",
    "        index_del = []  # remove results below the condition query_coverage > 90 & identity > 90.\n",
    "        for i in range(len(q_id)):\n",
    "            if (float(q_cov[i]) > 90) and (float(ident[i]) > 90): pass\n",
    "            else: index_del.append(i)\n",
    "        \n",
    "        # 'del' is applied immediately to remove the value corresponding to the index, so if you remove it without reversing the order, the next index value before the erase is skipped.\n",
    "        index_del.sort(reverse=True)\n",
    "\n",
    "        for i in index_del:\n",
    "            del q_id[i]\n",
    "            del s_id[i]\n",
    "            del q_cov[i]\n",
    "            del ident[i]\n",
    "\n",
    "        q_id_count = {}  # q_id Enter the number of each element\n",
    "        for i in set(q_id): q_id_count[i] = q_id.count(i)\n",
    "\n",
    "        # blast_sheet_{} (dict)\n",
    "        # : generate dynamic variables with matching results between query (target organism) and subject (others) (considering parallelog)\n",
    "        a = globals()['blast_sheet_{}'.format(serial)] = {}\n",
    "        for i in set(q_id):\n",
    "            if q_id_count[i] == 1: a[i] = s_id[q_id.index(i)]\n",
    "            else:\n",
    "                lst = []\n",
    "                for j in range(q_id.index(i), q_id.index(i) + q_id_count[i]): lst.append(s_id[j])\n",
    "                a[i] = lst\n",
    "\n",
    "        serial += 1\n",
    "\n",
    "        \n",
    "    print ('<DONE>\\nblast_orthologous({0})\\n'.format(x))\n",
    "        \n",
    "        \n",
    "\n",
    "# Final orthologous data trimming function\n",
    "# : If you have both EDGAR and BLAST+ orthologous data, this function integrate the two data (priority : EDGAR > BLAST+).\n",
    "# : Although you have only one orthologous data, this function trim orthologous data.\n",
    "def make_total_orthologous (file_name_ortho_edgar, file_name_ortho_blastp):\n",
    "    \n",
    "    print ('<RUNNING>\\nmake_total_orthologous()\\n: Final trimming of orthologous data...\\n')\n",
    "    \n",
    "    edgar_wb = openpyxl.load_workbook('{0}/input/{1}'.format(path, file_name_ortho_edgar))\n",
    "    edgar_sheet_names = edgar_wb.sheetnames\n",
    "    edgar = edgar_wb[edgar_sheet_names[0]]\n",
    "    \n",
    "    # if EDGAR 3.0 data, run this script.---------\n",
    "    edgar.delete_rows(1)\n",
    "\n",
    "    for i, j in enumerate(edgar['1']):  # max_column of EDGAR\n",
    "        if bool(j.value) == False:\n",
    "            edgar_sample_num = i\n",
    "            break\n",
    "\n",
    "    blast_wb = openpyxl.load_workbook('{0}/input/{1}'.format(path, file_name_ortho_blastp))\n",
    "    blast_sheet_names = blast_wb.sheetnames\n",
    "    blast_sample_num = len(blast_sheet_names)  # The number of pairwise-alignment on BLAST+\n",
    "\n",
    "\n",
    "    # Load dynamic variables of EDGAR and BLAST+ into str type\n",
    "    edgar_result_id = []\n",
    "    blast_result_id = []\n",
    "    for i in range(edgar_sample_num-1): edgar_result_id.append('edgar_A_vs_{}'.format(apb[i + 1]))\n",
    "    for i in range(blast_sample_num): blast_result_id.append('blast_sheet_{}'.format(i+1))\n",
    "\n",
    "    result_id = list(set(edgar_result_id) | set(blast_result_id))\n",
    "    result_id.sort()\n",
    "\n",
    "    \n",
    "    # generate a complete list of keys for the locus_tag of the target organism\n",
    "    total_keys = set()\n",
    "    for i in result_id:\n",
    "        a = globals()[i]\n",
    "        a_keys = set(a.keys())\n",
    "        total_keys = total_keys | a_keys\n",
    "    total_keys = list(total_keys)\n",
    "\n",
    "\n",
    "    # Combine EDGAR and BLAST+ into one (priority : EDGAR > BLAST+)\n",
    "    total_orthologous = {}\n",
    "    edgar_orthologous_result = {}\n",
    "    blast_orthologous_result = {}\n",
    "    for i in total_keys:\n",
    "        lst_bla = []\n",
    "        lst_ed = []\n",
    "        for j in result_id:\n",
    "            a = globals()[j]\n",
    "            if 'blast_sheet_' in j:\n",
    "                if i in a:\n",
    "                    if str(type(a[i])) == \"<class 'str'>\":\n",
    "                        lst_bla.append(a[i])\n",
    "                    else:\n",
    "                        lst_bla.extend(a[i])\n",
    "            elif 'edgar_A_vs_' in j:\n",
    "                if i in a:\n",
    "                    if str(type(a[i])) == \"<class 'str'>\":\n",
    "                        lst_ed.append(a[i])\n",
    "                    else:\n",
    "                        lst_ed.extend(a[i])\n",
    "        lst_bla = list(set(lst_bla))\n",
    "        lst_bla.sort()\n",
    "        if '-' in lst_bla:\n",
    "            lst_bla.remove('-')\n",
    "        blast_orthologous_result[i] = lst_bla\n",
    "        if lst_bla == []:\n",
    "            del blast_orthologous_result[i]\n",
    "        lst_ed = list(set(lst_ed))\n",
    "        lst_ed.sort()\n",
    "        if '-' in lst_ed:\n",
    "            lst_ed.remove('-')\n",
    "        edgar_orthologous_result[i] = lst_ed\n",
    "        if lst_ed == []:\n",
    "            del edgar_orthologous_result[i]\n",
    "\n",
    "    # put BLAST+ in total_orthologous (empty dict) first and then overwrite EDGAR.\n",
    "    # (If the dict of EDGAR and BLAST+ has the same key (target organism locus tag), select EDGAR.)\n",
    "    total_orthologous.update(blast_orthologous_result)\n",
    "    total_orthologous.update(edgar_orthologous_result)\n",
    "    \n",
    "    print ('<DONE>\\nmake_total_orthologous()\\n')\n",
    "\n",
    "\n",
    "    return total_orthologous"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Model reconstruction\n",
    "\n",
    "**1) Add reactions to draft model from reference models along with Gene-Protein-Reaction (GPR) relationships**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bring the model and combine the reactions that correspond to the orthologous.\n",
    "# args : SBML file path , (If there are multiple models, write them like 'A, B, C'.)\n",
    "# file_name_model_objective: file name of model with desired objective function\n",
    "# geme_remain_lst (optional): List of genes to keep even if they are not orthologous (e.g., virtual gene of spontaneous reaction)\n",
    "# our (optional): Oxygen uptake rate of draft model\n",
    "def make_draft_model (args, file_name_model_objective, gene_remain_lst = False, our = False):\n",
    "    \n",
    "    print ('<RUNNING>\\nmake_draft_model([{0}])\\n: Generating draft model from reference model...\\n'.format(', '.join(args)))            \n",
    "    \n",
    "    draft_model = cobra.Model('draft_model')  # Create an empty model\n",
    "    biomass_rxns = {}\n",
    "    model_path_lst = []\n",
    "    add_rxns_num_lst = []\n",
    "    for models in args:\n",
    "        model_ori = cobra.io.read_sbml_model('{0}/input/{1}'.format(path, models))\n",
    "        model = model_ori.copy()\n",
    "        model_path_lst.append('{0}/input/{1}'.format(path, models))\n",
    "        \n",
    "        if models == file_name_model_objective:\n",
    "            model_id_objective = model.id\n",
    "\n",
    "        model_genes = [] \n",
    "        for i in model.genes: model_genes.append(i.id)\n",
    "\n",
    "        com_genes = []  # overlaped genes extraction between model and othologous data\n",
    "        for k,v in total_orthologous.items():\n",
    "            for j in v:\n",
    "                if j in model_genes:\n",
    "                    com_genes.append(j)\n",
    "        remove_genes = list(set(model_genes) - set(com_genes))  # remove_genes : genes without target organism and othologous\n",
    "        remove_genes.sort()\n",
    "        \n",
    "        \n",
    "        # (optional) Remove the elements of gene_remain_lst from remove_genes to leave the elements of gene_remain_lst in the draft model.\n",
    "        if gene_remain_lst:        \n",
    "            for gene in remove_genes:\n",
    "                if gene in gene_remain_lst:\n",
    "                    remove_genes.remove(gene)\n",
    "        \n",
    "        \n",
    "        # Remove_genes from within the model. (Also remove the reactions that are deactivated without the corresponding genes in accordance with GPR.)\n",
    "        cobra.manipulation.delete.remove_genes(model, remove_genes)\n",
    "\n",
    "\n",
    "        # Add remaining reactions to the draft model\n",
    "        add_rxns = []\n",
    "        for i in model.reactions:\n",
    "            add_rxns.append(i)\n",
    "\n",
    "        pre_objective_id = str(model.objective.expression)\n",
    "        objective_id = pre_objective_id[pre_objective_id.index('*')+1 : pre_objective_id.index(' - ')]\n",
    "\n",
    "        objective_rxn = model.reactions.get_by_id(objective_id).copy()\n",
    "\n",
    "        for i in add_rxns:\n",
    "            if i.id == objective_id:\n",
    "                add_rxns.remove(i)\n",
    "\n",
    "        draft_model.add_reactions(add_rxns)\n",
    "        if model.id == model_id_objective:\n",
    "            add_rxns_num_lst.append(len(add_rxns) + 1)\n",
    "        else:\n",
    "            add_rxns_num_lst.append(len(add_rxns))\n",
    "\n",
    "        \n",
    "        # Save the objective function of each reference model as a dict type (key : model ID, value : objective function)\n",
    "        biomass_rxns[model.id] = objective_rxn\n",
    "\n",
    "\n",
    "    delay = 0\n",
    "    for i in range(10**7): # 'Ignoring reaction ...' Because input is in the middle of the warning, delay is required.\n",
    "        delay = delay + i - i\n",
    "        str(delay)\n",
    "        int(delay)\n",
    "\n",
    "\n",
    "    # Enable input for objective function setting\n",
    "    draft_model.add_reactions([biomass_rxns[model_id_objective]])\n",
    "    draft_model.objective = biomass_rxns[model_id_objective].id\n",
    "    objective_function_id = biomass_rxns[model_id_objective].id\n",
    "\n",
    "    print ('draft_model_objective_function: ' + objective_function_id)\n",
    "\n",
    "\n",
    "    # (optional) Set the upper limit of the oxygen uptake rate in the model\n",
    "    if our:\n",
    "        medium = draft_model.medium\n",
    "        medium['EX_o2_e'] = our\n",
    "        draft_model.medium = medium\n",
    "        print (draft_model.medium)\n",
    "        print ('EX_o2_e:', abs(draft_model.reactions.get_by_id('EX_o2_e').lower_bound))\n",
    "\n",
    "\n",
    "    print('=' * 20 + '\\n' + '{}'.format(draft_model.id) + '\\n' + '=' * 20)\n",
    "    print('reactions:', len(draft_model.reactions))\n",
    "    print('genes:', len(draft_model.genes))\n",
    "    print('metabolites:', len(draft_model.metabolites))\n",
    "    \n",
    "    \n",
    "    print ('Add reactions number')\n",
    "    for i in range(len(args)):\n",
    "        print ('Add reactions from {0}: {1}'.format(args[i], add_rxns_num_lst[i]))\n",
    "\n",
    "    \n",
    "    print ('<DONE>\\nmake_draft_model({0})\\n'.format(', '.join(args)))\n",
    "    \n",
    "\n",
    "    return draft_model\n",
    "\n",
    "\n",
    "# Change the gene locus tag of the draft model to target organism\n",
    "def rename (model, total_orthologous):  # model : target model\n",
    "    \n",
    "    print ('<RUNNING>\\nrename({0}, total_orthologous)\\n: Changing the gene ID of the draft model to locus tag of the target organism...\\n'.format(model))\n",
    "    \n",
    "    # Generating a list of genes in a model.\n",
    "    model_genes = []\n",
    "    model_genes_id = []\n",
    "    for i in model.genes:\n",
    "        model_genes.append(i)\n",
    "        model_genes_id.append(i.id)\n",
    "\n",
    "    # Find the paralog and set it according to gene_reaction_rule format\n",
    "    rename_dict = {}\n",
    "    for i in model_genes_id:\n",
    "        para_list = []\n",
    "        for k,v in total_orthologous.items():\n",
    "            for j in v:\n",
    "                if j == i:\n",
    "                    para_list.append(k)\n",
    "        if len(para_list) > 1:\n",
    "            rename_dict[i] = ['(' + ' or '.join(para_list) + ')', para_list]  # rename_dict[i][0] = gene_reaction_rule format, rename_dict[i][1] = gene_list\n",
    "        elif len(para_list) == 1:\n",
    "            rename_dict[i] = [para_list[0], para_list]\n",
    "\n",
    "\n",
    "    # Change genes ID of model to locus tag of target organism\n",
    "    for i in model.reactions:\n",
    "        gene_id_lst = []\n",
    "        for j in list(i.genes):\n",
    "            gene_id_lst.append(j.id)\n",
    "        for j in gene_id_lst:\n",
    "            if j in rename_dict:\n",
    "                new_gene_id_lst = []\n",
    "                for h in i.genes:\n",
    "                    new_gene_id_lst.append(h.id)\n",
    "                if len(set(rename_dict[j][1]) & set(new_gene_id_lst)) == 0:\n",
    "                    model.reactions.get_by_id(i.id).gene_reaction_rule = i.gene_reaction_rule.replace(j, rename_dict[j][0])\n",
    "\n",
    "    \n",
    "    print ('<DONE>\\nrename({0}, total_orthologous)\\n'.format(model))\n",
    "    \n",
    "    \n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# * Optional Functions\n",
    "- change_locus_tag\n",
    " : For EDGAR Server, Genome data exists as RefSeq DB's. So if the model's base Genome data is GenBank DB (old locus tag), then we need to replace RefSeq DB (locus tag) with GenBank DB (old locus tag).\n",
    " \n",
    "- match_lt_old_new (sub-function of change_locus_tag)\n",
    " : Function that extracts CDSs with both locus tag and old locus tag from genome file and matches locus tag with old locus tag.\n",
    "\n",
    "- output_excel_sbml_file\n",
    " : Function that outputs the model to EXCEL and SBML format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that extracts CDSs with both locus tag and old locus tag from genome file and matches locus tag with old locus tag\n",
    "def match_lt_old_new(x):  # x : genome file path. (e.g. A.gb)\n",
    "    \n",
    "    print ('<RUNNING>\\nmatch_lt_old_new({0})\\n: Matching locus tag and old locus tag from CDS type of genome...\\n'.format(x))\n",
    "    \n",
    "    file = next(SeqIO.parse(x, \"genbank\"))\n",
    "\n",
    "    feat = file.features\n",
    "\n",
    "    genome = []\n",
    "    for i in feat:\n",
    "        if i.qualifiers.get(\"pseudo\") == [\"\"]:\n",
    "            pass\n",
    "        elif i.type == 'CDS':\n",
    "            genome.append(i)\n",
    "\n",
    "    cds_old_lt = {}\n",
    "    for i in genome:\n",
    "        if 'old_locus_tag' in i.qualifiers:\n",
    "            cds_old_lt[i.qualifiers['locus_tag'][0]] = i.qualifiers['old_locus_tag'][0]\n",
    "\n",
    "    \n",
    "    print ('<DONE>\\nmatch_lt_old_new({0})\\n'.format(x))\n",
    "    \n",
    "            \n",
    "    return cds_old_lt\n",
    "\n",
    "\n",
    "\n",
    "# Function that changes the locus tag of a gene to an old locus tag\n",
    "# args : Genome file of RefSeq DB based on the same sequencing data\n",
    "# file_name_ortho_edgar: file name of EDGAR orthologous data\n",
    "def change_locus_tag (args, file_name_ortho_edgar):\n",
    "    \n",
    "    print ('<RUNNING>\\nchange_locus_tag({0})\\n: Changing to locus tag for the model of the genome...\\n'.format(args))\n",
    "    \n",
    "    edgar_wb = openpyxl.load_workbook('{0}/input/{1}'.format(path, file_name_ortho_edgar))\n",
    "    sheet_names = edgar_wb.sheetnames\n",
    "    edgar = edgar_wb[sheet_names[0]]\n",
    "\n",
    "    # if EDGAR 3.0 data, run this script.---------\n",
    "    edgar.delete_rows(1)\n",
    "    \n",
    "    \n",
    "    for i, j in enumerate(edgar['1']): # max column of EDGAR file\n",
    "        if bool(j.value) == False:\n",
    "            max_col = i\n",
    "            break\n",
    "\n",
    "    for genome_file in args:\n",
    "        cds_old_lt = match_lt_old_new('{0}/input/{1}'.format(path, genome_file))\n",
    "        cds_old_lt_keys = list(cds_old_lt.keys())\n",
    "        for i,j in enumerate(cds_old_lt_keys[0]):\n",
    "            try:\n",
    "                int(j)\n",
    "                num_index = i\n",
    "                break\n",
    "            except:\n",
    "                pass\n",
    "        target_chr = cds_old_lt_keys[0][: num_index]\n",
    "        target_chr\n",
    "        \n",
    "        print ('target_chr:', target_chr)\n",
    "\n",
    "        a_name_index = 0\n",
    "        for i in range(max_col-1):\n",
    "            a = globals()['edgar_A_vs_{}'.format(apb[i + 1])]\n",
    "            a_values = list(a.values())\n",
    "            \n",
    "            for h,j in enumerate(a_values):\n",
    "                if j != '-':\n",
    "                    a_values_index = h\n",
    "                    break\n",
    "            \n",
    "            for k,j in enumerate(a_values[a_values_index]):\n",
    "                try:\n",
    "                    int(j)\n",
    "                    num_index = k\n",
    "                    break\n",
    "                except:\n",
    "                    pass\n",
    "            a_chr = a_values[a_values_index][: num_index]\n",
    "            \n",
    "            print ('a_chr:', a_chr)\n",
    "            \n",
    "            if a_chr == target_chr:\n",
    "                \n",
    "                print ('chrs:', a_chr, target_chr)\n",
    "                \n",
    "                a_name_index += i\n",
    "                break\n",
    "        \n",
    "        print (a_name_index)\n",
    "\n",
    "        a_copy = globals()['edgar_A_vs_{}'.format(apb[a_name_index + 1])].copy()\n",
    "        globals()['edgar_A_vs_{}'.format(apb[a_name_index + 1])].clear()\n",
    "        for k, v in a_copy.items():\n",
    "            if str(type(v)) == \"<class 'str'>\":\n",
    "                if v in cds_old_lt:\n",
    "                    globals()['edgar_A_vs_{}'.format(apb[a_name_index + 1])][k] = cds_old_lt[v]\n",
    "            else:\n",
    "                lst = []\n",
    "                for j in v:\n",
    "                    if j in cds_old_lt:\n",
    "                        lst.append(cds_old_lt[j])\n",
    "                if len(lst) == 0:\n",
    "                    pass\n",
    "                elif len(lst) == 1:\n",
    "                    globals()['edgar_A_vs_{}'.format(apb[a_name_index + 1])][k] = cds_old_lt[v[0]]\n",
    "                elif len(lst) > 1:\n",
    "                    globals()['edgar_A_vs_{}'.format(apb[a_name_index + 1])][k] = lst\n",
    "    \n",
    "    \n",
    "    print ('<DONE>\\nchange_locus_tag({0})\\n'.format(args))\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "# Function to output the model to SBML and EXCEL format\n",
    "def output_excel_sbml_file(model, model_id):  # model : target model, model_id : Storage name and model name of the output file of the model\n",
    "    \n",
    "    print ('<RUNNING>\\noutput_excel_sbml_file({0}, {1})\\n: Outputing model to excel and sbml format...\\n'.format(model, model_id))\n",
    "    \n",
    "    model.id = model_id\n",
    "    \n",
    "    cobra.io.write_sbml_model(model, '{0}/output/{1}.xml'.format(path, model.id))\n",
    "    \n",
    "    model_rxn_info = {}\n",
    "    model_rxn_id = []\n",
    "    for i in model.reactions:\n",
    "        model_rxn_id.append(i.id)\n",
    "        model_rxn_info[i.id] = [i.id , i.name, i.reaction, i.gene_reaction_rule, i.lower_bound, i.upper_bound, i.objective_coefficient, i.subsystem]\n",
    "\n",
    "    model_mt_info = {}\n",
    "    model_mt_id = []\n",
    "    for i in model.metabolites:\n",
    "        model_mt_id.append(i.id)\n",
    "        comp = ''\n",
    "        if i.compartment == 'c':\n",
    "            comp = 'Cytosol'\n",
    "        elif i.compartment == 'p':\n",
    "            comp = 'Periplasm'\n",
    "        elif i.compartment == 'e':\n",
    "            comp = 'Extracellular'\n",
    "\n",
    "        model_mt_info[i.id] = [i.id , i.name, i.formula, i.charge, comp]\n",
    "\n",
    "    header_rxn = ['Reaction ID', 'Description', 'Reaction', 'Gene-protein-reaction (GPR) rules', 'Lower bound', 'Upper bound', 'Objective', 'Subsystem']\n",
    "\n",
    "    header_mt = ['Metabolite ID', 'Description', 'Charged formula', 'Charge', 'Compartment']\n",
    "\n",
    "\n",
    "    write_wb = openpyxl.Workbook()\n",
    "    write_ws_1 = write_wb.create_sheet('Reaction List')\n",
    "    write_ws_2 = write_wb.create_sheet('Metabolite List')\n",
    "    write_wb.remove(write_wb['Sheet'])\n",
    "\n",
    "    write_ws_1 = write_wb['Reaction List']\n",
    "    write_ws_1.append(header_rxn)\n",
    "    for i in model_rxn_id:\n",
    "        write_ws_1.append(model_rxn_info[i])\n",
    "\n",
    "    write_ws_2 = write_wb['Metabolite List']\n",
    "    write_ws_2.append(header_mt)\n",
    "    for i in model_mt_id:\n",
    "        write_ws_2.append(model_mt_info[i])\n",
    "\n",
    "    write_wb.save('{0}/output/{1}.xlsx'.format(path, model.id))\n",
    "    \n",
    "    \n",
    "    print ('<DONE>\\noutput_excel_sbml_file({0}, {1})\\n'.format(model, model_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# * Running Script\n",
    "**- precondition**\n",
    " - The genome file of target organism must exist.\n",
    " - At least one reference model must exist.\n",
    " - There must be an othologous file compared between the genome files and the target organism based on the reference model construction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example\n",
    ": make draft model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### trimming orthologous file\n",
    "\n",
    "# target_organism_genome_file\n",
    "file_name_target_gb = 'NZ_CP009974.1.gb'\n",
    "target_cds_lt = extract_cds_lt_in_genome ('{0}/input/{1}'.format(path, file_name_target_gb))\n",
    "\n",
    "# edgar_orthologous file\n",
    "file_name_ortho_edgar = 'edgar_ortho.xlsx'\n",
    "edgar_orthologous ('{0}/input/{1}'.format(path, file_name_ortho_edgar))\n",
    "\n",
    "# blastp_orthologous file\n",
    "file_name_ortho_blastp = 'blastp_ortho.xlsx'\n",
    "blast_orthologous ('{0}/input/{1}'.format(path, file_name_ortho_blastp))\n",
    "\n",
    "# (optional) Change the locus tag of orthologous data to the old locus tag of the model\n",
    "file_name_change_gb_lst = ['NC_002947.4.gb']\n",
    "change_locus_tag(file_name_change_gb_lst, file_name_ortho_edgar)\n",
    "\n",
    "# orthologous_file_total_trimming\n",
    "total_orthologous = make_total_orthologous(file_name_ortho_edgar, file_name_ortho_blastp)  # integrity of edgar files and blast files into one (priority : EDGAR > BLAST+)\n",
    "\n",
    "\n",
    "### reconstructing model file\n",
    "\n",
    "# input Reference models\n",
    "file_name_model_lst = ['iJN1463.xml', 'iPAE1146.xml'] # write model file names in order of highest priority.\n",
    "file_name_model_objective = 'iJN1463.xml'\n",
    "gene_remain_lst = ['PP_s0001','SPONTANEOUS','PA2366']\n",
    "our = 18.5\n",
    "draft_model = make_draft_model(file_name_model_lst, file_name_model_objective, gene_remain_lst, our)\n",
    "\n",
    "# rename in draft_model\n",
    "draft_model = rename(draft_model, total_orthologous)\n",
    "\n",
    "# (optional) export draft_model\n",
    "output_excel_sbml_file (draft_model, model_id = 'draft_model')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
